### PR TITLE
Asim/kustoking/add watchguard parser

### PR DESCRIPTION
**Pull request by KustoKing**
Required items, please complete
   
   Change(s):
   - Add Watch Guard Fireware OS parser

   Reason for Change(s):
   - New parser

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    | extend SrcDomainType = iif(isnotempty(SrcDomain),"FQDN","")
+    ;
+    let Parse1 = 
+    AllSyslog
+    | where NetworkProtocol !in("icmp","igmp") and SyslogMessage !has "3000-0151"
+    | parse kind=regex flags=U SyslogMessage with * @'" (Allow|Deny) ' RuleName @" \d{2,5} (tcp|udp) \d{2,5} \d{2,5} " SrcIpAddr " " DstIpAddr " " SrcPortNumber:int @" " DstPortNumber:int @" " *
+    | where  (isnull(dstportnumber) or DstPortNumber==dstportnumber)
+    | invoke IPParser()
+    | invoke HostParser()
+    | invoke SyslogParser()
```
Comment (oshezaf) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml:140]: As a reminder, the reason to have those invokes was to add a project-away in each branch.

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      | extend DstUsernameType = iif(DstUsername has "@","UPN","")
+      | parse SyslogMessage with * "repeated " EventCount:int " times" *
+      | extend EventCount = iif(isnotempty(EventCount), EventCount, toint(1))
+      | project-away SyslogMessage
+    };
+    let IPParser = (Syslog:(SrcIpAddr:string,DstIpAddr:string)){
+      Syslog
+      | extend temp_SrcMatch = has_any_ipv4_prefix(SrcIpAddr,src_or_any)
+      , temp_DstMatch = has_any_ipv4_prefix(DstIpAddr,dst_or_any)
+      | extend ASimMatchingIpAddr = case(
```
Comment (oshezaf) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml:88]: Did you test the filters? It has to be manual and is very error prone. We can discuss tomorrow how to best do that.

**Review by oshezaf — CHANGES_REQUESTED** [review]

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    | parse kind=regex flags=U SyslogMessage with * @'" (Allow|Deny) ' RuleName @" icmp " SrcIpAddr " " DstIpAddr " " * 
+    | invoke SyslogParser()
+    ;
+    let Parse5 = 
+    AllSyslog
+    | where NetworkProtocol == "igmp" and SyslogMessage !has "3000-0151"
+    | parse kind=regex flags=U SyslogMessage with * @'" (Allow|Deny) ' RuleName @" \d{2,5} igmp \d{2,5} \d{1,5} " SrcIpAddr " " DstIpAddr " " * 
+    | invoke SyslogParser()
+    ;
+    union Parse1, Parse2, Parse3, Parse4, Parse5
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:99]: Ok

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    Default: false
+ParserQuery: |
+  let Parser=(disabled:bool=false){
+    let EventLookup=datatable(DvcAction:string,EventResult:string,EventSeverity:string)
+    [
+        "Allow","Success","Informational"
+        , "Deny","Failure","Low"
+    ];
+    let SyslogParser = (Syslog:(SyslogMessage:string)) {
+        Syslog
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:31]: Ok

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+      , NetworkDuration = duration
+      , SrcBytes = sent_bytes
+      , DstBytes = rcvd_bytes
+      | extend SrcUsernameType = iif(SrcUsername has "@","UPN","")
+      | extend DstUsernameType = iif(DstUsername has "@","UPN","")
+      | parse SyslogMessage with * "repeated " EventCount:int " times" *
+      | extend EventCount = iif(isnotempty(EventCount), EventCount, toint(1))
+      | project-away SyslogMessage
+    };
+    let IPParser = (Syslog:(SrcIpAddr:string,DstIpAddr:string)){
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml:84]: Ok

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+Description: |
+  This ASIM parser supports normalizing WatchGuard Fireware OS logs to the ASIM Network Session normalized schema.
+ParserName: ASimNetworkSessionWatchGuardFirewareOS
+EquivalentBuiltInParser: _ASim_NetworkSession_WatchGuardFirewareOS
+ParserParams:
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
+  let Parser=(disabled:bool=false){
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:24]: Cleaned columns in AllSyslog

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        , sent_bytes:long
+        , rcvd_bytes:long) with (pair_delimiter=' ', kv_delimiter='=', quote='"')
+        | project-rename SrcGeoCountry = geo_src
+        , DstGeoCountry = geo_dst
+        , SrcUsername = src_user
+        , DstUsername = dst_user
+        , NetworkDuration = duration
+        , SrcBytes = sent_bytes
+        , DstBytes = rcvd_bytes
+        | extend SrcUsernameType = iif(SrcUsername has "@","UPN","")
```
[the remark this answers is not in the document]
Comment (oshezaf) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:46]: Use case to cover for empty as well.

[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:46]: Ok

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    | extend DvcAction = extract(@'" (Allow|Deny) ', 1, SyslogMessage)
+    | lookup EventLookup on DvcAction
+    | extend NetworkProtocol = extract(@" (tcp|udp|icmp|igmp) ", 1, SyslogMessage)
+    | parse-kv SyslogMessage as (fqdn_src_match:string, fqdn_dst_match:string) with (pair_delimiter=' ', kv_delimiter='=', quote='"')
+    | project-rename DstDomain = fqdn_dst_match, SrcDomain = fqdn_src_match
+    | extend DstDomainType = iif(isnotempty(DstDomain),"FQDN","")
+    | extend SrcDomainType = iif(isnotempty(SrcDomain),"FQDN","")
+    ;
+    let Parse1 = 
+    AllSyslog
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionWatchGuardFirewareOS.yaml:79]: OK

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml [hunk 9]
```diff
@@ -0,0 +1,196 @@
+Parser:
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml:1]: OK

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml (hunk 2 above)
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml:101]: Filters are tested and work.

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml (hunk 1 above)
Comment (KustoKing) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionWatchGuardFirewareOS.yaml]: From AllSyslog, I project only the required fields, then from the Syslog Parser I project away the SyslogMessage

**Review by oshezaf — APPROVED** [review]

**Pull request merged**